core/src/main/scala/kafka/metrics/ClientMetricsMetadata.scala

```diff
@@ -0,0 +1,119 @@
+ * limitations under the License.
+ */
+package kafka.metrics
+
+import kafka.metrics.ClientMetricsConfig.ClientMatchingParams._
+import kafka.network.RequestChannel
+import org.apache.kafka.common.errors.InvalidConfigurationException
+
+import java.util.regex.{Pattern, PatternSyntaxException}
+import scala.collection.mutable
+
+/**
+ * Information from the client's metadata is gathered from the client's request.
+ */
+object ClientMetricsMetadata {
+  def apply(request: RequestChannel.Request, clientInstanceId: String): ClientMetricsMetadata = {
+    val instance = new ClientMetricsMetadata
+    val ctx = request.context
+    val clientSoftwareName = if (ctx.clientInformation != null) ctx.clientInformation.softwareName() else ""
+    val clientSoftwareVersion = if (ctx.clientInformation != null) ctx.clientInformation.softwareVersion() else ""
+    instance.init(clientInstanceId, ctx.clientId(), clientSoftwareName, clientSoftwareVersion,
+                  ctx.clientAddress.getHostAddress, ctx.clientAddress.getHostAddress) // TODO: Fix Port
+    instance
+  }
+
+  def apply(clientInstanceId: String, clientId: String, clientSoftwareName: String,
+            clientSoftwareVersion: String, clientSourceAddress: String, clientSourcePort: String): ClientMetricsMetadata = {
+    val instance = new ClientMetricsMetadata
+    instance.init(clientInstanceId, clientId, clientSoftwareName, clientSoftwareVersion, clientSourceAddress, clientSourcePort)
+    instance
+  }
+
+  /**
+   * Parses the client matching patterns and builds a map with entries that has
+   * (PatternName, PatternValue) as the entries.
+   *  Ex: "VERSION=1.2.3" would be converted to a map entry of (Version, 1.2.3)
+   *
+   *  NOTES:
+   *  1. Client match pattern splits the input into two parts separated by first
+   *     occurrence of the character '='
+   *  2. '*' is considered as invalid client match pattern
+   * @param patterns List of client matching pattern strings
+   * @return map of client matching pattern entries
+   */
+  def parseMatchingPatterns(patterns: List[String]) : Map[String, String] = {
+    val patternsMap = mutable.Map[String, String]()
+    if (patterns != null) {
+      patterns.foreach(x => {
+        val nameValuePair = x.split("=", 2).map(x => x.trim)
+        if (nameValuePair.size == 2 && isValidParam(nameValuePair(0)) && validRegExPattern(nameValuePair(1))) {
+          patternsMap += (nameValuePair(0) -> nameValuePair(1))
+        } else {
+          throw new InvalidConfigurationException("Illegal client matching pattern: " + x)
+        }
+      })
+    }
+    patternsMap.toMap
+  }
+
+  private def validRegExPattern(inputPattern :String): Boolean = {
+    try {
+      Pattern.compile(inputPattern)
+      true
+    } catch {
+      case _: PatternSyntaxException =>
+        false
+    }
+  }
+
+}
+
+class ClientMetricsMetadata {
+  var attributesMap: mutable.Map[String, String] = scala.collection.mutable.Map[String, String]()
```

Comment:
Do you need the init? If you don't, you can just do this:
```
private val attributesMap: mutable.Map[String, String] = mutable.Map(
    CLIENT_INSTANCE_ID -> clientInstanceId,
    CLIENT_ID -> clientId,
    CLIENT_SOFTWARE_NAME -> clientSoftwareName,
    CLIENT_SOFTWARE_VERSION -> clientSoftwareVersion,
    CLIENT_SOURCE_ADDRESS -> clientSourceAddress,
    CLIENT_SOURCE_PORT -> clientSourcePort
  )
```

such that the apply method above can just be 

```
def apply(...): ClientMetricsMetadata = {
    new ClientMetricsMetadata(...)
  }
```

Reply:
Valid, I ll take it up once decided to continue this class in scala or move to respective java package.

Reply:
Moved to java implementation